File: 2024/Day_09/part01.py
import numpy as np

# reads in data from file
with open("test.txt", "r") as file:
    data = np.array(list(file.read()), dtype=int)
    # adds 0 to last file with NO partitions after
    if len(data) % 2 != 0:
        data = np.append(data, 0)

# create partition map of harddrive
def harddrive_map(paired_files):
    harddrive = []
    id_val = 0

    for pair in paired_files:
        harddrive.extend([id_val] * pair[0])
        harddrive.extend([-1] * pair[1])
        id_val += 1

    return harddrive

# shifts far right files to furthest empty left
def optimize_space(harddrive):
    opt_harddrive = []
    left_ptr = 0
    right_ptr = len(harddrive) - 1

    while (left_ptr <= right_ptr):
        if (harddrive[left_ptr] != -1):
            opt_harddrive.append(harddrive[left_ptr])
        # empty partition
        else:
            # find furthest right file (non empty)
            while harddrive[right_ptr] == -1: 
                right_ptr -= 1
            if (left_ptr < right_ptr):
                opt_harddrive.append(harddrive[right_ptr])
            right_ptr -= 1
        left_ptr += 1
        
    return opt_harddrive

def calculate_checksum(opt_harddrive):
    checksum = 0
    for idx, val in enumerate(opt_harddrive):
        checksum += idx * val
    return checksum

paired_files = data.reshape(-1, 2)          # pairs 1st and 2nd elements together
harddrive = harddrive_map(paired_files)
opt_harddrive = optimize_space(harddrive)
checksum = calculate_checksum(opt_harddrive)
print(checksum)

File: 2024/Day_09/part01.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
using namespace std;

vector<int> read_file() {
    ifstream file("test.txt");
    if (!file.is_open()) {
        cerr << "Error Opening File" << endl;
        exit(-1);
    }

    string line;
    vector<int> data;
    while(getline(file, line)) {
        for (char ch : line)
            data.push_back(ch - '0');  // char to int
    }

    if (data.size() % 2 != 0) data.push_back(0);
    file.close();
    return data;
}

vector<pair<int, int>> make_pairs(vector<int> &data) {
    vector<pair<int, int>> pairedFiles;

    for (size_t i = 0; i < data.size(); i += 2) {
        pairedFiles.push_back({data[i], data[i + 1]});
    }

    return pairedFiles;
}

vector<int> harddrive_map(vector<pair<int, int>> &pairedFiles) {
    vector<int> harddrive;
    int idVal = 0;

    for (int i = 0; i < pairedFiles.size(); i++) {
        harddrive.insert(harddrive.end(), pairedFiles[i].first, idVal);
        harddrive.insert(harddrive.end(), pairedFiles[i].second, -1);
        idVal++;
    }

    return harddrive;
}

vector<int> optimize_space(vector<int> &harddrive) {
    vector<int> optHarddrive;
    int leftPtr = 0;
    int rightPtr = harddrive.size()-1;

    while (leftPtr <= rightPtr) {
        if (harddrive[leftPtr] != -1)
            optHarddrive.push_back(harddrive[leftPtr]);
        else {
            while (harddrive[rightPtr] == -1)
                rightPtr -= 1;
            if (leftPtr < rightPtr)
                optHarddrive.push_back(harddrive[rightPtr]);
            rightPtr--;
        }
        leftPtr++;
    }
        
    return optHarddrive;
}

long long calculate_checksum(vector<int> &optHarddrive) {
    long long checksum = 0;
    for (int i = 0; i < optHarddrive.size(); i++) {
        checksum += (i * optHarddrive[i]);
    }
    return checksum;
}



int main() {
    vector<int> data = read_file();
    vector<pair<int, int>> pairedFiles = make_pairs(data);
    vector<int> harddrive = harddrive_map(pairedFiles);
    vector<int> optHarddrive = optimize_space(harddrive);
    long long checksum = calculate_checksum(optHarddrive);
    cout << checksum << endl;
    return 0;
}